CRM/Financial/BAO/Payment.php

```diff
@@ -560,4 +572,61 @@ protected static function isPaymentCompletesContribution($contributionID, $payme
     return ($cmp == 0 || $cmp == 1);
   }
 
+  /**
+   * Get the line items for the contribution.
+   *
+   * Retrieve the line items and wrangle the following
+   *
+   * - get the outstanding balance on a line item basis.
+   * - determine what amount is being paid on this line item - we get the total being paid
+   *   for the whole contribution and determine the ratio of the balance that is being paid
+   *   and then assign apply that ratio to each line item.
+   * - if overrides have been passed in we use those amounts instead.
+   *
+   * @param $params
+   *
+   * @return array
+   * @throws \CiviCRM_API3_Exception
+   */
+  protected static function getPayableLineItems($params): array {
+    $lineItems = CRM_Price_BAO_LineItem::getLineItemsByContributionID($params['contribution_id']);
+    $lineItemOverrides = CRM_Utils_Array::value('line_item', $params, []);
+    $outstandingBalance = CRM_Contribute_BAO_Contribution::getContributionBalance($params['contribution_id']);
+    if ($outstandingBalance !== 0) {
+      $ratio = $params['total_amount'] / $outstandingBalance;
+    }
+    else {
+      // Help we are making a payment but no money is owed. We won't allocate the overpayment to any line item.
+      $ratio = 0;
+    }
+    foreach ($lineItems as $lineItemID => $lineItem) {
+      $lineItems[$lineItemID]['paid'] = 0;
+      $financialItems = civicrm_api3('FinancialItem', 'get', [
+        'entity_id' => $lineItemID,
+        'entity_table' => 'civicrm_line_item',
+        'options' => ['sort' => 'id DESC'],
+      ])['values'];
+      if (!empty($financialItems)) {
+        $entityFinancialTrxns = civicrm_api3('EntityFinancialTrxn', 'get', [
+          'entity_table' => 'civicrm_financial_item',
+          'entity_id' => ['IN' => array_keys($financialItems)],
+        ])['values'];
```

Comment:
@eileenmcnaughton not all data stored in EntityFinancialTrxn are stored as paid some are pending incase of 'Accounts receivable' so probably you may need to join civicrm_financial_trxn table to check is_payment = 1 (i.e 'financial_trxn_id.is_payment' => 1)

Reply:
Do we need to set limit => 0?

Reply:
Yes - well spotted - fixed

Reply:
hmm. is_payment is also set true for refunds. I'm going to play around with output on different use cases.